test/e2e/clusterctl_upgrade.go

```diff
@@ -0,0 +1,300 @@
+			KubernetesVersion:        kubernetesVersion,
+			ControlPlaneMachineCount: controlPlaneMachineCount,
+			WorkerMachineCount:       workerMachineCount,
+			InfrastructureProvider:   clusterctl.DefaultInfrastructureProvider,
+			// setup clusterctl logs folder
+			LogFolder: filepath.Join(input.ArtifactFolder, "clusters", managementClusterProxy.GetName()),
+		})
+		Expect(workloadClusterTemplate).ToNot(BeNil(), "Failed to get the cluster template")
+
+		log.Logf("Applying the cluster template yaml to the cluster")
+		Expect(managementClusterProxy.Apply(ctx, workloadClusterTemplate)).To(Succeed())
+
+		By("Waiting for the machines to exists")
+		Eventually(func() (int64, error) {
+			var n int64 = 0
+			machineList := &clusterv1old.MachineList{}
+			if err := managementClusterProxy.GetClient().List(ctx, machineList, client.InNamespace(testNamespace.Name), client.MatchingLabels{clusterv1.ClusterLabelName: clusterName}); err == nil {
+				for _, machine := range machineList.Items {
+					if machine.Status.NodeRef != nil {
+						n++
+					}
+				}
+			}
+			return n, nil
+		}, input.E2EConfig.GetIntervals(specName, "wait-worker-nodes")...).Should(Equal(*controlPlaneMachineCount + *workerMachineCount))
+
+		By("THE MANAGEMENT CLUSTER WITH OLDER VERSION OF PROVIDERS WORKS!")
+
+		By("Upgrading providers to the latest version available")
+		clusterctl.UpgradeManagementClusterAndWait(ctx, clusterctl.UpgradeManagementClusterAndWaitInput{
+			ClusterctlConfigPath: input.ClusterctlConfigPath,
+			ClusterProxy:         managementClusterProxy,
+			ManagementGroup:      fmt.Sprintf("capi-system/%s", config.ClusterAPIProviderName),
+			Contract:             clusterv1.GroupVersion.Version,
+			LogFolder:            filepath.Join(input.ArtifactFolder, "clusters", cluster.Name),
+		}, input.E2EConfig.GetIntervals(specName, "wait-controllers")...)
+
+		By("THE MANAGEMENT CLUSTER WAS SUCCESSFULLY UPGRADED!")
+
+		// After upgrading we are sure the version is the latest version of the API,
+		// so it is possible to use the standard helpers
+
+		testMachineDeployments := framework.GetMachineDeploymentsByCluster(ctx, framework.GetMachineDeploymentsByClusterInput{
+			Lister:      managementClusterProxy.GetClient(),
+			ClusterName: clusterName,
+			Namespace:   testNamespace.Name,
+		})
+
+		framework.ScaleAndWaitMachineDeployment(ctx, framework.ScaleAndWaitMachineDeploymentInput{
+			ClusterProxy:              managementClusterProxy,
+			Cluster:                   &clusterv1.Cluster{ObjectMeta: metav1.ObjectMeta{Namespace: testNamespace.Name}},
+			MachineDeployment:         testMachineDeployments[0],
+			Replicas:                  2,
+			WaitForMachineDeployments: input.E2EConfig.GetIntervals(specName, "wait-worker-nodes"),
+		})
+
+		By("THE UPGRADED MANAGEMENT CLUSTER WORKS!")
+
+		By("PASSED!")
+	})
+
+	AfterEach(func() {
+		if testNamespace != nil {
+			// Dump all the logs from the workload cluster before deleting them.
```

Comment:
Should we get the logs and dump all resources from the management and workload cluster?

Reply:
Ah thx, missed that somehow